### PR TITLE
Include Rotation class and unit tests

### PR DESCRIPTION
**Pull request by junchaoxia**
First PR for closing issue #195. The Rotation class and corresponding unit tests are created.

### CODE REVIEW

File: src/aspire/utils/rotation.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+from scipy.linalg import svd
+from scipy.spatial.transform import Rotation as sp_rot
+
+from aspire.utils import ensure
+from aspire.utils.matlab_compat import Random
+
+
+class Rotation:
+    """
+    Define a Customized Rotation class
```
Comment (vineetbansal) [inline src/aspire/utils/rotation.py]: No need for this docstring

File: src/aspire/utils/rotation.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    """
+
+    def __init__(self, num_rots, seed=None, seq='ZYZ',
+                 angles=None, degrees=False, dtype=np.float32):
+        """
+        Initialize a Rotation object
+
+        :param num_rots: Total number of rotation sets
+        :param seed: Seed for initializing random rotations
```
Comment (vineetbansal) [inline src/aspire/utils/rotation.py]: If None, then what?

File: src/aspire/utils/rotation.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+        :param seq:  Sequence of order applying Euler angles
+        :param angles: Euler angles to generate rotation matrices
+        :param degrees: Whether Euler angles in degrees or not
```
Comment (vineetbansal) [inline src/aspire/utils/rotation.py]: if False, then what?

File: src/aspire/utils/rotation.py [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+        :param dtype:   data type for angles and rotation matrices
+        """
+        self.rot_seq = seq
+        self.degrees = degrees
+        self.dtype = np.dtype(dtype)
+        if angles:
+            ensure(num_rots == angles.shape[0],
+                   'Number of rotation matrices should equal to '
+                   'number of sets of Euler angles.')
+            self.angles = angles
```
Comment (vineetbansal) [inline src/aspire/utils/rotation.py]: probably should have `.astype(self.dtype)` too?

File: src/aspire/utils/rotation.py [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+        """
+        self._rotations = sp_rot.from_matrix(values)
+        self.data = values.astyp(self.dtype)
+
+    @property
+    def angles(self):
+        """
+        :return: Rotation angles in radians, as a n x 3 array
+        """
+        return self._rotations.as_euler(self.rot_seq)
```
Comment (vineetbansal) [inline src/aspire/utils/rotation.py]: If `self.degree`, return in degrees.

File: src/aspire/utils/rotation.py [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+    @angles.setter
+    def angles(self, values):
```
Comment (vineetbansal) [inline src/aspire/utils/rotation.py]: If `self.degree`, assume input is in degrees.

File: src/aspire/utils/rotation.py [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+    @rot_matrices.setter
+    def rot_matrices(self, values):
+        """
+        Set rotation matrices
+
+        :param values: Rotation matrices as a n x 3 x 3 array
+        :return: None
+        """
+        self._rotations = sp_rot.from_matrix(values)
+        self.data = values.astyp(self.dtype)
```
Comment (vineetbansal) [inline src/aspire/utils/rotation.py]: astype

File: tests/test_rotation.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        print(rots_ref[1])
+        print(self.rot_obj.rot_matrices[1])
+        self.assertTrue(np.allclose(self.rot_obj.rot_matrices, rots_ref))
+
+    def testRotAngles(self):
+        rot = sp_rot.from_matrix(self.rot_obj.rot_matrices)
+        angles = rot.as_euler(self.rot_obj.rot_seq)
+        self.assertTrue(np.allclose(self.rot_obj.angles, angles))
+
+    def TestTranspose(self):
```
Comment (vineetbansal) [inline tests/test_rotation.py]: lowecase t

File: tests/test_rotation.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        rot_mat = self.rot_obj.rot_matrices
+        rot_mat_t = self.rot_obj.T
+        for i in range(self.rot_obj.num_rots):
+            self.assertTrue(np.allclose(rot_mat_t[i], rot_mat.T[i]))
+
+    def TestMultiplication(self):
```
Comment (vineetbansal) [inline tests/test_rotation.py]: lowercase t

File: tests/test_rotation.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        rot_obj_t =  Rotation(32, seed=0, seq='ZYZ')
+        rot_obj_t.rot_matrices = self.rot_obj.T
+        result = np.allclose(self.rot_obj*rot_obj_t)
```
Comment (vineetbansal) [inline tests/test_rotation.py]: a bug that will be revealed when lowercasing the t

File: src/aspire/utils/rotation.py (hunk 2 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: The random numbers are generated using the customized  Random class in aspire.utils.matlab_compat. If seed is None. it will continue from the previous random state. More notations were added in new commit 9f6e3b3.

File: src/aspire/utils/rotation.py (hunk 1 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Removed in new commit 9f6e3b3.

File: src/aspire/utils/rotation.py (hunk 3 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Since our application cases  in `ImageSource` class use degrees, I have removed this option and well will use degrees directly in new commit 9f6e3b3.

File: src/aspire/utils/rotation.py (hunk 4 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done.

File: src/aspire/utils/rotation.py (hunk 5 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Changed to use  degrees in default.

File: src/aspire/utils/rotation.py (hunk 6 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Changed to use degrees in default.

File: src/aspire/utils/rotation.py (hunk 7 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Fixed.

File: tests/test_rotation.py (hunk 8 above)
Comment (junchaoxia) [reply tests/test_rotation.py]: Done.

File: tests/test_rotation.py (hunk 9 above)
Comment (junchaoxia) [reply tests/test_rotation.py]: Done.

File: tests/test_rotation.py (hunk 10 above)
Comment (junchaoxia) [reply tests/test_rotation.py]: Done in  commit 4836a01.

File: src/aspire/utils/rotation.py (hunk 2 above)
Comment (vineetbansal) [reply src/aspire/utils/rotation.py]: Cool..maybe say "If None, the random generator is not re-seeded."

File: tests/test_rotation.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
+from unittest import TestCase
+import numpy as np
+
+from scipy.spatial.transform import Rotation as sp_rot
+from aspire.utils.rotation import Rotation
+from aspire.utils.types import utest_tolerance
+
+
+class UtilsTestCase(TestCase):
+    def setUp(self):
```
Comment (vineetbansal) [inline tests/test_rotation.py:10]: You don't need 3 attributes here - just have `self.rot_obj` and look inside it to get what you want.

File: src/aspire/utils/rotation.py [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+            T[i] = self.data[i].T
+        return T
+
+    def __str__(self):
+        """
+        String representation.
+        """
+        return f'Rotation object with matrices[{self.num_rots}, 3, 3] of {self.dtype} type'
+
+    def copy(self):
```
Comment (vineetbansal) [inline src/aspire/utils/rotation.py]: This should be `__copy__`, but then it will lead to inf. recursion. Better to not have this at all.

File: src/aspire/utils/rotation.py [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+
+    def __getitem__(self, item):
+        return self.data[item]
+
+    def __setitem__(self, key, value):
+        self.data[key] = value
+
+    def __mul__(self, other):
+        if isinstance(other, Rotation):
+            other = other.data
```
Comment (vineetbansal) [inline src/aspire/utils/rotation.py]: output = self.data @ other.data

File: src/aspire/utils/rotation.py [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+        """
+        return self.transpose()
+
+    def transpose(self):
+        """
+        Apply transpose operation to all rotation matrices
+
+        :return: The set of transposed matrices
+        """
+        T = np.zeros((self.num_rots, 3, 3), dtype=self.dtype)
```
Comment (vineetbansal) [inline src/aspire/utils/rotation.py]: np.transpose(self.data, axes=(0, 2, 1))

File: src/aspire/utils/rotation.py [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+        for i in range(self.num_rots):
+            T[i] = self.data[i].T
+        return T
+
+    def __str__(self):
```
Comment (vineetbansal) [inline src/aspire/utils/rotation.py]: move all the dunder methods towards the top (after __init__). All staticmethods should be clustered together, and either go before regular instance methods, or after (see the rest of the code to see if there's a convention we're following).

File: tests/test_rotation.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        self.dtype = np.float32
+        self.rot_obj = Rotation(32, seed=0, seq='ZYZ', dtype=self.dtype)
```
Comment (vineetbansal) [inline tests/test_rotation.py]: Maybe get rid of the 'ZYZ' when calling the constructor in both places (since one of the motivations of this class was to hide that stuff under the rug).

File: src/aspire/utils/rotation.py (hunk 2 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done in new commit 9786c56.

File: src/aspire/utils/rotation.py (hunk 12 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Removed in new commit  9786c56.

File: src/aspire/utils/rotation.py (hunk 13 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done in new commit  9786c56.

File: src/aspire/utils/rotation.py (hunk 14 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done in new commit  9786c56.

File: src/aspire/utils/rotation.py (hunk 15 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done in new commit  9786c56.

File: tests/test_rotation.py (hunk 11 above)
Comment (junchaoxia) [reply tests/test_rotation.py:10]: Removed in commit 707aace.

File: tests/test_rotation.py (hunk 16 above)
Comment (junchaoxia) [reply tests/test_rotation.py]: Removed in commit 707aace.

File: src/aspire/utils/rotation.py (hunk 13 above)
Comment (vineetbansal) [reply src/aspire/utils/rotation.py]: The `other = other.data` line needs to go

File: tests/test_rotation.py (hunk 16 above)
Comment (vineetbansal) [reply tests/test_rotation.py]: `testMultiplication` still has this

File: src/aspire/utils/rotation.py (hunk 13 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done in commit 948928f.

File: tests/test_rotation.py (hunk 16 above)
Comment (junchaoxia) [reply tests/test_rotation.py]: Done in commit 948928f.

File: src/aspire/utils/rotation.py [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+from numpy.linalg import norm
+from scipy.linalg import svd
+from scipy.spatial.transform import Rotation as sp_rot
+
+from aspire.utils import ensure
+from aspire.utils.random import Random
+
+
+class Rotation:
+    def __init__(self, num_rots, seed=None, seq="ZYZ", angles=None, dtype=np.float32):
```
Comment (janden) [inline src/aspire/utils/rotation.py]: It's not a good idea to have the default constructor generate random rotations. It would be better for it to accept angles or rotation matrices and then have a static method that initializes a `Rotation` object with uniformly distributed random rotations.

File: src/aspire/utils/rotation.py [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+                num_rots == angles.shape[0],
+                "Number of rotation matrices should equal to "
+                "number of sets of Euler angles.",
+            )
+            self.angles = angles.astype(self.dtype)
+        else:
+            self.angles = self._uniform_random_angles(num_rots, seed=seed)
+        self.num_rots = num_rots
+
+        self.data = self.rot_matrices
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Why are these needed? Isn't this data already contained elsewhere?

File: src/aspire/utils/rotation.py [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+        self.shape = (num_rots, 3, 3)
+
+    def __str__(self):
+        """
+        String representation.
+        """
+        return (
+            f"Rotation object with matrices[{self.num_rots}, 3, 3] of {self.dtype} type"
```
Comment (janden) [inline src/aspire/utils/rotation.py]: No need to mention matrices. Just "Rotation stack consisting of {self.num_rots} elements of {self.dtype} type".

File: src/aspire/utils/rotation.py [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+        Set rotation angles in degrees
+
+        :param values: Rotation angles in degrees, as a n x 3 array
+        :return: None
+        """
+        self._rotations = sp_rot.from_euler(
+            self.rot_seq, values.astype(self.dtype), degrees=True
+        )
+
+    @property
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Remove.

File: src/aspire/utils/rotation.py [hunk 21]
```diff
@@ -0,0 +1,280 @@
...
+    def T(self):
+        """
+        Shortcut to get full set of transposed rotation matrices
+        """
+        return self.transpose()
+
+    def transpose(self):
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Better to call it `invert`.

File: src/aspire/utils/rotation.py [hunk 22]
```diff
@@ -0,0 +1,280 @@
...
+                    np.random.random(n) * 2 * np.pi,
+                    np.arccos(2 * np.random.random(n) - 1),
+                    np.random.random(n) * 2 * np.pi,
+                )
+            )
+        # Return random rotation angles in degrees
+        return np.rad2deg(angles).astype(self.dtype)
+
+    @property
+    def rot_matrices(self):
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Why not just `matrices`? Besides, isn't this just available through the `data` attribute?

File: src/aspire/utils/rotation.py [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+        with Random(seed):
+            angles = np.column_stack(
+                (
+                    np.random.random(n) * 2 * np.pi,
+                    np.arccos(2 * np.random.random(n) - 1),
+                    np.random.random(n) * 2 * np.pi,
+                )
+            )
+        # Return random rotation angles in degrees
+        return np.rad2deg(angles).astype(self.dtype)
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Let's stick to radians as much as possible.

File: src/aspire/utils/rotation.py [hunk 24]
```diff
@@ -0,0 +1,280 @@
...
+        return self.data[item]
+
+    def __setitem__(self, key, value):
+        self.data[key] = value
+
+    def __mul__(self, other):
+        if isinstance(other, Rotation):
+            output = self.data @ other.data
+        else:
+            output = self.data * other
```
Comment (janden) [inline src/aspire/utils/rotation.py]: When is this supposed to be used?

File: src/aspire/utils/rotation.py [hunk 25]
```diff
@@ -0,0 +1,280 @@
...
+        # Reflection matrix
+        J = np.array([[1, 0, 0], [0, 1, 0], [0, 0, -1]])
+
+        regrot = np.zeros_like(rots)
+        for k in range(K):
+            R = rots[k, :, :]
+            if flag == 1:
+                R = J @ R @ J
+            regrot[k, :, :] = Q_mat.T @ R
+        aligned_rots = copy.copy(self)
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Why copy? Just create a new `Rotation` object.

File: src/aspire/utils/rotation.py [hunk 26]
```diff
@@ -0,0 +1,280 @@
...
+
+    def transpose(self):
+        """
+        Apply transpose operation to all rotation matrices
+
+        :return: The set of transposed matrices
+        """
+        return np.transpose(self.data, axes=(0, 2, 1))
+
+    def asnumpy(self):
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Isn't this the same as `rot_matrices` above?

File: src/aspire/utils/rotation.py [hunk 27]
```diff
@@ -0,0 +1,280 @@
...
+        return self.data
+
+    def register_rotations(self, rots_ref):
```
Comment (janden) [inline src/aspire/utils/rotation.py]: `find_registration`

File: src/aspire/utils/rotation.py [hunk 28]
```diff
@@ -0,0 +1,280 @@
...
+        else:
+            # Use Q2 as the registering matrix
+            U, _, V = svd(Q2)
+            flag = 1
+
+        Q_mat = U @ V
+
+        return Q_mat, flag
+
+    def get_aligned_rotations(self, Q_mat, flag):
```
Comment (janden) [inline src/aspire/utils/rotation.py]: `apply_registration` and create another method `register` that combines `find_registration` and `apply_registration`.

File: src/aspire/utils/rotation.py [hunk 29]
```diff
@@ -0,0 +1,280 @@
...
+            R = rots[k, :, :]
+            if flag == 1:
+                R = J @ R @ J
+            regrot[k, :, :] = Q_mat.T @ R
+        aligned_rots = copy.copy(self)
+        aligned_rots.rot_matrices = regrot
+        return aligned_rots
+
+    @staticmethod
+    def get_rots_mse(rots_reg, rots_ref):
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Why not just make this a method called `mse`?

File: src/aspire/utils/rotation.py [hunk 30]
```diff
@@ -0,0 +1,280 @@
...
+        diff = np.zeros(K)
+        mse = 0
+        for k in range(K):
+            diff[k] = norm(rots_reg[k, :, :] - rots_ref[k, :, :], ord="fro")
+            mse += diff[k] ** 2
+        mse = mse / K
+        return mse
+
+    @staticmethod
+    def common_line_from_rots(r1, r2, ell):
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Again, why not make this a method `common_lines`?

File: tests/test_rotation.py [hunk 31]
```diff
@@ -0,0 +1,52 @@
...
+        result = self.rot_obj * rot_obj_t
+        for i in range(self.rot_obj.num_rots):
+            self.assertTrue(
+                np.allclose(
+                    np.eye(3), result[i], atol=utest_tolerance(self.rot_obj.dtype)
+                )
+            )
+
+    def testRegisterRots(self):
+        q_ang = [[45, 45, 45]]
```
Comment (janden) [inline tests/test_rotation.py]: Pick some random angles.

File: tests/test_rotation.py [hunk 32]
```diff
@@ -0,0 +1,52 @@
...
+        q_mat = sp_rot.from_euler("ZYZ", q_ang, degrees=True).as_matrix()[0]
+        flag = 0
```
Comment (janden) [inline tests/test_rotation.py]: Test for both flags.

**Review by janden — CHANGES_REQUESTED** [review]
Let's discuss at the meeting today.

File: src/aspire/utils/rotation.py (hunk 17 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Changed to initialize by angles or matrices in commit 0c8f0b8.

File: src/aspire/utils/rotation.py (hunk 18 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Intended to make consistent with other classes by using `data` as internal usage and `rot_matrices` for external interface. Changed to use `matrices` for both of them in new commit  0c8f0b8.

File: src/aspire/utils/rotation.py (hunk 19 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done.

File: src/aspire/utils/rotation.py (hunk 20 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Removed in commit 0c8f0b8.

File: src/aspire/utils/rotation.py (hunk 21 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Renamed in commit ac3f350.

File: src/aspire/utils/rotation.py (hunk 22 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Combined them into one in commit 0c8f0b8.

File: src/aspire/utils/rotation.py (hunk 23 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Changed to use radians in commit 0c8f0b8.

File: src/aspire/utils/rotation.py (hunk 24 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Tried to make it work when np array is passed instead of Rotation object. It should be `output = self.data @ other`. Changed in commit ac3f350.

File: src/aspire/utils/rotation.py (hunk 25 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Changed to return a new `Rotation` object in commit e863d01.

File: src/aspire/utils/rotation.py (hunk 26 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Tried to make a consistent interface with Image and Volume class which can return numpy array. `rot_matrices` was replaced by  `matrices` in new commit.

File: src/aspire/utils/rotation.py (hunk 27 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done in commit e863d01.

File: src/aspire/utils/rotation.py (hunk 28 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done in commit e863d01.

File: src/aspire/utils/rotation.py (hunk 29 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done in commit e863d01.

File: src/aspire/utils/rotation.py (hunk 30 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: The original thinking is get common lines between two rotation matrices which have no relation with current object. Changed to input two indices of matrices stored in the current Rotation object. We might change it later when coming to the detailed implementation  in orientation methods.

File: tests/test_rotation.py (hunk 31 above)
Comment (junchaoxia) [reply tests/test_rotation.py]: Done in new commit e863d01.

File: tests/test_rotation.py (hunk 32 above)
Comment (junchaoxia) [reply tests/test_rotation.py]: Done in commit e863d01.

File: src/aspire/utils/rotation.py [hunk 33]
```diff
@@ -0,0 +1,300 @@
...
+        """
+        Generate random 3D rotation angles in radians
+
+        :param n: The number of rotation angles to generate
+        :param seed: Random integer seed to use. If None,
+            the current random state is used.
+        :return: A n-by-3 ndarray of rotation angles in radians
+        """
+        # Generate random rotation angles, in radians
+        angles = np.zeros((n, 3), dtype=dtype)
```
Comment (vineetbansal) [inline src/aspire/utils/rotation.py]: This line is not used. When you remove this, add `dtype` to the return. Add docstring for dtype.

File: src/aspire/utils/rotation.py (hunk 17 above)
Comment (janden) [reply src/aspire/utils/rotation.py]: Pick one: initialize from angles *or* matrices. Since a lot of the interface suggests that this is a matrix-like object, it would make more sense to initialize from matrices.

Also, it still generates random angles if neither are supplied. Remove this.

File: src/aspire/utils/rotation.py (hunk 20 above)
Comment (janden) [reply src/aspire/utils/rotation.py]: No still there.

File: src/aspire/utils/rotation.py [hunk 34]
```diff
@@ -0,0 +1,300 @@
...
+         :param angles: Euler angles (in radian) to generate rotation matrices.
+             If None, Predefined rotation matrices or uniformly
+             distributed angles will be used.
+        :param matrices: Rotation matrices to initialize Rotation object.
+             If None, Predefined or uniformly distributed angles
+             will be used.
+         :param dtype: data type for angles and rotation matrices
+        """
+        self.dtype = np.dtype(dtype)
+        self.num_rots = None
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Why do we need to store `num_rots`, `angles`, and `shape` separately? They can all be obtained from `matrices`.

File: src/aspire/utils/rotation.py [hunk 35]
```diff
@@ -0,0 +1,300 @@
...
+
+    def __setitem__(self, key, value):
+        self.matrices[key] = value
+
+    def __mul__(self, other):
+        if isinstance(other, Rotation):
+            output = self.matrices @ other.matrices
+        else:
+            output = self.matrices @ other
+        return output
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Shouldn't this return another `Rotation` instance?

File: src/aspire/utils/rotation.py [hunk 36]
```diff
@@ -0,0 +1,300 @@
...
+        """
+        return self.invert()
+
+    def invert(self):
+        """
+        Apply transpose operation to all rotation matrices
+
+        :return: The set of transposed matrices
+        """
+        return np.transpose(self.matrices, axes=(0, 2, 1))
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Again, why not a `Rotation` instance?

File: src/aspire/utils/rotation.py [hunk 37]
```diff
@@ -0,0 +1,300 @@
...
+
+    def asnumpy(self):
+        """
+        :return: Rotation matrices as a n x 3 x 3 array
+        """
+        return self.matrices
+
+    def from_euler(self, values, dtype=np.float32):
```
Comment (janden) [inline src/aspire/utils/rotation.py]: The two `from_` functions should be static.

File: src/aspire/utils/rotation.py [hunk 38]
```diff
@@ -0,0 +1,300 @@
...
+        """
+        build rotation object from Euler angles in degrees
+
+        :param dtype:  data type for rotational angles and matrices
+        :param values: Rotation angles in radians, as a n x 3 array
+        :return: None
+        """
+        self.dtype = np.dtype(dtype)
+        self.num_rots = values.shape[0]
+        rotations = sp_rot.from_euler("ZYZ", values.astype(dtype), degrees=False)
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Hardcode `"ZYZ"` into some private attribute so we don't accidentally change it somewhere.

File: src/aspire/utils/rotation.py [hunk 39]
```diff
@@ -0,0 +1,300 @@
...
+
+    def asnumpy(self):
+        """
+        :return: Rotation matrices as a n x 3 x 3 array
+        """
+        return self.matrices
+
+    def from_euler(self, values, dtype=np.float32):
+        """
+        build rotation object from Euler angles in degrees
```
Comment (janden) [inline src/aspire/utils/rotation.py]: “radians”

File: src/aspire/utils/rotation.py (hunk 24 above)
Comment (janden) [reply src/aspire/utils/rotation.py]: Yes but when is it supposed to be used? When do we want to multiply a Rotation by an `np.array` instance?

File: src/aspire/utils/rotation.py [hunk 40]
```diff
@@ -0,0 +1,300 @@
...
+
+
+class Rotation:
+    def __init__(
+        self, num_rots, seed=None, angles=None, matrices=None, dtype=np.float32
+    ):
+        """
+         Initialize a Rotation object
+
+         :param num_rots: Total number of rotation sets
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Why does this need to be specified separately from `matrices`?

File: src/aspire/utils/rotation.py [hunk 41]
```diff
@@ -0,0 +1,300 @@
...
+
+    def invert(self):
+        """
+        Apply transpose operation to all rotation matrices
+
+        :return: The set of transposed matrices
+        """
+        return np.transpose(self.matrices, axes=(0, 2, 1))
+
+    def asnumpy(self):
```
Comment (janden) [inline src/aspire/utils/rotation.py]: If this is the preferred way to access the array from the outside, let's make `matrices` into `_matrices` to indicate that it's private.

File: tests/test_rotation.py (hunk 31 above)
Comment (janden) [reply tests/test_rotation.py]: Why not use `uniform_random_angles` for this?

File: src/aspire/utils/rotation.py [hunk 42]
```diff
@@ -0,0 +1,300 @@
...
+        ell_ij = alpha_ij * ell / (2 * np.pi)
+        ell_ji = alpha_ji * ell / (2 * np.pi)
+
+        ell_ij = int(np.mod(np.round(ell_ij), ell))
+        ell_ji = int(np.mod(np.round(ell_ji), ell))
+
+        return ell_ij, ell_ji
+
+    @staticmethod
+    def uniform_random_angles(
```
Comment (janden) [inline src/aspire/utils/rotation.py]: `generate_random_rotations` and have it return a `Rotation` instance.

**Review by janden — CHANGES_REQUESTED** [review]

File: src/aspire/utils/rotation.py (hunk 17 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: In real case we might have three major applications. 1) Generate Rotation object for Simulation class using random angles:  2) Generate Rotation object using rotational angles from STAR file; 3) Generate Rotation object for the rotation matrices predicted from orientation estimation. The constructor tries to cover these three cases. The using of angles is convenient so that no transform outside the class is needed?

File: src/aspire/utils/rotation.py (hunk 20 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: I removed all @property  except for .T which is used to represent transpose matrices instead of a function .T() in a way similar to numpy object or BlkDiagMatrix object. Remove too?

File: src/aspire/utils/rotation.py (hunk 24 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Removed and simplified in new commit  8cbe5e1.

File: tests/test_rotation.py (hunk 31 above)
Comment (junchaoxia) [reply tests/test_rotation.py]: Changed in new commit 2602b07.

File: src/aspire/utils/rotation.py (hunk 33 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done.

File: src/aspire/utils/rotation.py (hunk 34 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: When we don't specify angles and matrices such as the case of using random angles, we need num_rots to set the number of rotation matrices.

File: src/aspire/utils/rotation.py (hunk 35 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Changed to return Rotation object in commit 466729d.

File: src/aspire/utils/rotation.py (hunk 36 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Changed to return Rotation object in commit 466729d.

File: src/aspire/utils/rotation.py (hunk 37 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Changed to use static property in new commit a6c2f53.

File: src/aspire/utils/rotation.py (hunk 38 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Set the sequence order to a private attribute in commit 8ad8fb8, used to be a public one with default value.

File: src/aspire/utils/rotation.py (hunk 39 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done.

File: src/aspire/utils/rotation.py (hunk 40 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: For the case of using random angles in `Simulation` object where both `angles` and `matrices` are None.

File: src/aspire/utils/rotation.py (hunk 41 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: The `asnumpy` was removed  3c701b9.

File: src/aspire/utils/rotation.py (hunk 42 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: The `generate_random_rotations` was added as a static method in commit a82b980. The `uniform_random_angles` was kept for cases we would like to have angles.

File: src/aspire/utils/rotation.py (hunk 17 above)
Comment (janden) [reply src/aspire/utils/rotation.py]: The constructor should do *one thing*. I suggest that one thing should be “build a rotation stack from an array of rotation matrices.” If we want other ways to construct a rotation stack, we use the static methods. Note that this is why the SciPy rotation object creation looks the way it does (there you aren't even supposed to use the constructor).

File: src/aspire/utils/rotation.py (hunk 20 above)
Comment (janden) [reply src/aspire/utils/rotation.py]: Yes but transpose here is ambigious. Does it mean transpose the entire array to (3, 3, n) or does it mean to transpose each individual rotation matrix? The NumPy interpretation would imply the former. It's clearer to just have the method `invert` (you can't invert a (n, 3, 3) array).

Comment (janden) [reply src/aspire/utils/rotation.py]: So yes. Remove.

File: src/aspire/utils/rotation.py (hunk 24 above)
Comment (janden) [reply src/aspire/utils/rotation.py]: Ok. There's no need for the `isinstance` call now, though, right?

File: src/aspire/utils/rotation.py (hunk 41 above)
Comment (janden) [reply src/aspire/utils/rotation.py]: Why remove it? It's generally better to allow access to the underlying data through a getter method than making users access internal attributes.

File: src/aspire/utils/rotation.py (hunk 42 above)
Comment (janden) [reply src/aspire/utils/rotation.py]: When do we need to have angles? Shouldn't we be using a `Rotation` object in those instances?

File: src/aspire/utils/rotation.py (hunk 17 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Remove `angles` from the constructor and use `_matrices` to hold rotation stack in new commit f7fcc5cf.

File: src/aspire/utils/rotation.py (hunk 20 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: `T ` was removed in commit  0a4768d.

File: src/aspire/utils/rotation.py (hunk 41 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Changed the public variable `matrices` to private `_matrices` and added  `matrices` as a property for external use in commit ed13087.

File: src/aspire/utils/rotation.py (hunk 42 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: The `uniform_random_angles` was removed in commit  080a94e.

File: src/aspire/utils/rotation.py (hunk 17 above)
Comment (janden) [reply src/aspire/utils/rotation.py]: This is still generating random rotations if no matrices are supplied. The signature should be
```
def __init__(self, matrices)
```
that's it. All other ways to generate a Rotation object should be through the static methods.

Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done in new commit 51beb74.

File: src/aspire/utils/rotation.py [hunk 43]
```diff
@@ -12,39 +12,18 @@
...
-             If None, the random generator is not re-seeded.
         :param matrices: Rotation matrices to initialize Rotation object.
-             If None, matrices will be generated from uniformly
-             distributed angles.
-         :param dtype: data type for rotation matrices
         """
-        self.dtype = np.dtype(dtype)
-        self.num_rots = num_rots
-        self._matrices = None
+        self.dtype = matrices.dtype
```
Comment (janden) [inline src/aspire/utils/rotation.py]: Since this is always equal to `matrices.dtype`, let's make it a property so that we don't change it by accident.

File: src/aspire/utils/rotation.py [hunk 44]
```diff
@@ -12,39 +12,18 @@
...
+        self.num_rots = matrices.shape[0]
```
Comment (janden) [inline src/aspire/utils/rotation.py]: `num_rots` and `shape` are no longer necessary since these can be read out from `_matrices`.

**Review by janden — CHANGES_REQUESTED** [review]

File: src/aspire/utils/rotation.py (hunk 43 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Done in new commit ecfd5fd.

File: src/aspire/utils/rotation.py (hunk 44 above)
Comment (junchaoxia) [reply src/aspire/utils/rotation.py]: Removed in new commit ecfd5fd.

**Review by janden — DISMISSED** [review]

**Review by janden — APPROVED** [review]
